public/app/plugins/datasource/loki/formatterTS.ts

```diff
@@ -0,0 +1,406 @@
+import { SyntaxNode } from '@lezer/common';
+import { trimEnd } from 'lodash';
+
+import {
+  Eq,
+  Identifier,
+  String,
+  LogExpr,
+  Matcher,
+  parser,
+  PipelineExpr,
+  Selector,
+  Nre,
+  Re,
+  Neq,
+  LabelParser,
+  LabelFilter,
+  JsonExpressionParser,
+  LineFormatExpr,
+  LabelFormatExpr,
+  LineComment,
+  LineFilter,
+  Filter,
+  Regexp,
+  Pattern,
+  JsonExpression,
+  IpLabelFilter,
+  UnitFilter,
+  NumberFilter,
+  DurationFilter,
+  BytesFilter,
+  Gtr,
+  Gte,
+  Lss,
+  Lte,
+  Eql,
+  Duration,
+  Bytes,
+  Number,
+  LabelFormatMatcher,
+  FilterOp,
+} from '@grafana/lezer-logql';
+
+export const formatLogQL = (query: string): string => {
+  const tree = parser.parse(query);
+  let formatted = '';
+
+  tree.iterate({
+    enter: (ref): void => {
+      const node = ref.node;
+
+      switch (node.type.id) {
+        case Selector:
+          formatted += formatSelector(node, query);
+          break;
+
+        case PipelineExpr:
+          node.parent?.type.id === LogExpr && (formatted += formatPipelineExpr(node, query));
+          break;
+      }
+    },
+  });
+
+  return formatted;
+};
+
+export function formatSelector(node: SyntaxNode, query: string): string {
+  const selector = query.substring(node.from, node.to);
+  const subtree = parser.parse(selector);
+  const labelNodes: SyntaxNode[] = [];
+  let response = '';
+
+  // get all label nodes from the selector
+  subtree.iterate({
+    enter: (ref): void => {
+      const node = ref.node;
+      if (node.type.id === Matcher) {
+        labelNodes.push(node);
+      }
+    },
+  });
+
+  // sort the label nodes in alphabetical order
+  labelNodes.sort((a, b) => {
+    const LabelNodeA = a.getChild(Identifier)!;
+    const LabelNodeB = b.getChild(Identifier)!;
+
+    const labelValueA = query.substring(LabelNodeA.from, LabelNodeA.to);
+    const labelValueB = query.substring(LabelNodeB.from, LabelNodeB.to);
+
+    if (labelValueA < labelValueB) {
+      return -1;
+    }
+    if (labelValueA > labelValueB) {
+      return 1;
+    }
+    return 0;
+  });
+
+  // add the formatted label to the response
+  labelNodes.forEach((node) => {
+    const labelNode = node.getChild(Identifier)!;
```

Comment:
You should not use the non-null assertion at all, or as little as possible. There is a reason why TS wants you to check that `labelNode` is not actually null, and it's because it _can_ be null. A malformed query will cause all sort of errors and unexpected running conditions in your code, so even if it's annoying, the whole purpose of TS is to protect you, the code, and the user from these kind of situations.

![imagen](https://github.com/grafana/grafana/assets/1069378/967dd57b-7094-4a8b-a23c-cd6de5f8500a)